Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_from_build_products.py

```diff
@@ -0,0 +1,83 @@
+import argparse
+import os
+import shutil
+import stat
+import subprocess
+import tempfile
+from typing import cast
+from pathlib import Path
+
+
+def is_executable_file(path):
+    ext = os.path.splitext(path)[1]
+    mode = stat.S_IFREG | stat.S_IXUSR
+    return ext == '.a' or ext in ('', '.dylib') and \
+        os.stat(path).st_mode & mode == mode
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='wrapper around `tapi sdkdb` to produce a partial sdkdb '
+        'from Mach-O build products.')
+    parser.add_argument('products', nargs='+', type=Path,
+                        help='frameworks, dylibs, or root directories to '
+                        'include in generated sdkdb')
+    parser.add_argument('--sdk-dir',
+                        help='SDK to use when parsing public headers')
+    parser.add_argument('--output',
+                        help='path to write resulting .sdkdb to')
+    args = parser.parse_args()
+
+    sdk_dir = args.sdk_dir or os.environ['SDKROOT']
+    output_file = args.output or \
+        args.products[0].with_suffix('.partial.sdkdb').name
+
+    workdir = tempfile.TemporaryDirectory(prefix='sdkdb-from-build-products')
+    sdk_content = Path(workdir.name, 'SDKContentRoot')
+    runtime_content = Path(workdir.name, 'Root')
+    tapi_output = Path(workdir.name, 'sdkdb/partial.sdkdb')
+
+    for product in cast(list[Path], args.products):
+        parts = product.parts
+        for idx in reversed(range(len(parts))):
+            part = parts[idx]
+            if part.endswith('.framework'):
+                path_in_root = Path('System/Library/PrivateFrameworks',
+                                    *parts[idx:])
+                break
+            elif part.endswith('.dylib') or part.endswith('.a'):
+                path_in_root = Path('usr/local/lib', *parts[idx:])
+                break
+            elif part in ('usr', 'System'):
+                path_in_root = Path(*parts[idx:])
+                break
+        else:
+            raise RuntimeError("error: can't determine where to put"
+                               f"{product} in a synthetic root")
+        try:
+            shutil.copytree(product, sdk_content / path_in_root,
+                            ignore=lambda path, contents: [
+                                name for name in contents
+                                if is_executable_file(os.path.join(path, name))
+                            ])
+            shutil.copytree(product, runtime_content / path_in_root)
+        except NotADirectoryError:
```

Comment:
I'm not a fan of using exceptions for non-exceptional code paths, but I understand that Python directory copying is a mess and this could be the most elegant solution.